src/compiler/checker.ts

```diff
@@ -38905,9 +38905,14 @@ namespace ts {
         function unwrapReturnType(returnType: Type, functionFlags: FunctionFlags) {
             const isGenerator = !!(functionFlags & FunctionFlags.Generator);
             const isAsync = !!(functionFlags & FunctionFlags.Async);
-            return isGenerator ? getIterationTypeOfGeneratorFunctionReturnType(IterationTypeKind.Return, returnType, isAsync) || errorType :
-                isAsync ? getAwaitedTypeNoAlias(returnType) || errorType :
-                returnType;
+            if (isGenerator) {
+                const returnIterationType = getIterationTypeOfGeneratorFunctionReturnType(IterationTypeKind.Return, returnType, isAsync);
+                if (!returnIterationType) {
+                    return errorType;
+                }
+                return isAsync ? getAwaitedTypeNoAlias(unwrapAwaitedType(returnIterationType)) : returnIterationType;

```

Comment:
this is the core of the change: `isAsync` check + `unwrapAwaitedType`, the `getAwaitedTypeNoAlias` here seems to be irrelevant though (tests pass without it). I've found this pattern to be used in the codebase (`getAwaitedTypeNoAlias(unwrapAwaitedType(t))`) so I've decided to use it but we might want to either:
- remove the `getAwaitedTypeNoAlias` call
- figure out a test case that would showcase its relevance here